src/Plugin/Field/FieldWidget/ExploreCoursesUrlWidget.php

```diff
@@ -0,0 +1,171 @@
+      '#access' => TRUE,
+      '#element_validate' => [
+        [$this, 'validateApi'],
+      ],
+    ];
+    return $elements;
+  }
+
+  /**
+   * Validate we are specifying the newest api.
+   *
+   * @param array $element
+   *   Url form element.
+   * @param \Drupal\Core\Form\FormStateInterface $form_state
+   *   Current form state object.
+   * @param array $complete_form
+   *   Complete form.
+   */
+  public function validateApi(array &$element, FormStateInterface $form_state, array &$complete_form) {
+    $input = NestedArray::getValue($form_state->getValues(), $element['#parents']);
+
+    if ($form_state::hasAnyErrors()) {
+      return;
+    }
+    try {
+      $response = $this->client->request('GET', 'search?view=xml-' . $input, ['base_uri' => 'https://explorecourses.stanford.edu/']);
+      $response = (string) $response->getBody();
+      $xml = new \SimpleXMLElement($response);
+      // Do this as a string, since SimpleXMLElement doesn't cast to bools.
+      if ((string) $xml->deprecated == 'true') {
+        $form_state->setError($element, $this->t("That API version is deprecated. Newest version is: $xml->latestVersion"));
+      }
+    }
+    catch (\Throwable $e) {
+      $form_state->setError($element, $this->t('There was a problem querying the ExploreCourses API.'));
+    }
+  }
+
+  /**
+   * {@inheritDoc}.
+   */
+  public function settingsSummary() {
+    $summary = [];
+    if (empty($this->getSetting('api_version'))) {
+      $summary[] = $this->t('No API version Provided');
+    }
+    else {
+      $summary[] = $this->t('API version: @api_version', ['@api_version' => $this->getSetting('api_version')]);
+    }
+    return $summary;
+  }
+
+  /**
+   * {@inheritDoc}.
+   */
+  public function formElement(FieldItemListInterface $items, $delta, array $element, array &$form, FormStateInterface $form_state) {
+    $element = parent::formElement($items, $delta, $element, $form, $form_state);
+
+    $element['uri']['#access'] = TRUE;
+    $element['title']['#access'] = FALSE;
+    $element['attributes']['#access'] = FALSE;
+
+    return $element;
+  }
+
+  /**
+   * {@inheritDoc}.
+   */
+  public function massageFormValues(array $values, array $form, FormStateInterface $form_state) {
+
+    // Something like `view=xml-20200810`, but it may change.
+    $xml_querystring = 'xml-' . $this->getSetting('api_version');
+
+    foreach ($values as $delta => &$value) {
+      if (!empty($value['uri'])) {
+        // Parse the existing URL.
+        $url = UrlHelper::parse($value['uri']);
+
+        if (!str_contains($url['path'], 'explorecourses')) {
+          $form_state->setErrorByName("su_explore_course_url[$delta]", $this->t('The URL is not a valid ExploreCourses URL.'));
+          return;
+        }
+
+        // Check if the 'view' querystring doesn't exist, or if it's wrong.
+        if (empty($url['query']['view']) || empty($url['query']['view']) != '$xml_querystring') {
```

Comment:
This logic doesn't make sense. I think you're missing something like this:
```suggestion
        if (empty($url['query']['view']) || empty($url['query']['view']) || $url['query']['view'] != $xml_querystring) {
```

Reply:
Also you could just use `preg_replace` or something

Reply:
logic fixed.